Convert accessions to use a workflow model

ttung opened the pull request:
### Description
This turned out to be a lot more complicated than I expected.  Most of the complexity lies in the conversion logic (which we are **not** doing) and the logic for sample view.  The latter is required and now adds another roundtrip to the server.  I suspect it should be possible to do what we want in one query, but I can't seem to massage sqlalchemy to do what I want.

Rest of this is pretty straightforward.

#### Issue
[ch123630](https://app.clubhouse.io/genepi/story/123630/move-accessions-to-a-workflow-model)

### Test plan
make -j style unit-tests

phoenixAja commented on src/py/aspen/database/models/entity.py:
```diff
@@ -123,3 +123,29 @@ def get_children(
         return results
+
+    def accessions(self) -> Sequence[Accession]:
+        from .accessions import Accession
```
why add this import in the methods like this?

ttung replied:
Accessions imports Entity, so if I import Accessions from Entity, it'll be a recursive import.

phoenixAja replied:
do you think there's a way to reorganize the files so we don't have to do this? I find these workarounds for circular dependencies a little confusing.. if not maybe we can add this to our agenda to resolve later on?

ttung replied:
Not really. :(

phoenixAja commented on src/py/database_migrations/versions/20210330_120956_accessions_to_workflow_model.py:
```diff
@@ -0,0 +1,46 @@
+        sa.ForeignKeyConstraint(
+            ["workflow_id"],
+            ["aspen.workflows.id"],
+            name=op.f("fk_accession_workflows_workflow_id_workflows"),
```
so many workflow lol

phoenixAja commented on src/py/aspen/app/views/sample.py:
```diff
@@ -29,15 +24,21 @@ def _format_created_date(sample: Sample) -> str:
-        for accession in sample.sequencing_reads_collection.accessions:
-            if accession.repository_type == PublicRepositoryType.GISAID:
-                return accession.public_identifier
+        accession = entity_id_to_accession_map.get(
+            sample.sequencing_reads_collection.entity_id, None
+        )
+        if accession is not None:
+            return accession.public_identifier
```
this code is really similar to the above block, could probably create another function that returns the accession public_identifier?

ttung replied:
https://github.com/chanzuckerberg/aspen/pull/232

phoenixAja commented on src/py/aspen/covidhub_import/implementation.py:
```diff
@@ -301,19 +302,19 @@ def import_project(
 
             if czbid.genome_submission_info is not None:
+                repository_type: Optional[PublicRepositoryType] = None
+                public_identifier: str
```
why not just define this type hint below on line 309 when you set the `public_identifier` variable?

ttung replied:
i can omit this line if you don't like it.  it's not really that important, as the line on 309 will infer type correctly.

ttung replied:
Removed.

phoenixAja commented on src/py/aspen/database/models/accessions.py:
```diff
@@ -24,23 +26,52 @@ class PublicRepositoryType(enum.Enum):
-    entity = relationship(Entity, backref=backref("accessions", uselist=True))
+    entity_id = Column(Integer, ForeignKey(Entity.id), primary_key=True)
 
+    # statistics for the pathogen genome
```
is this comment correct? `repository_type` doesn't seem like it would refer to statistics

ttung replied:
nope, this is wrong.

ttung replied:
removed.

phoenixAja commented on src/py/aspen/database/models/entity.py:
```diff
@@ -123,3 +123,29 @@ def get_children(
+        """Adds an accession to this object."""
+        from .accessions import Accession
+
+        Accession.add_to_entity(
+            self,
+            repository_type,
+            public_identifier,
+            workflow_start_datetime,
+            workflow_end_datetime,
+        )
```
i also find this a little confusing (passing self as the entity for add_to_entity), and the whole entity/ accession relationship, because i read this like add an accession for this entity and then add the accession to this entity. Maybe it's the end of the day and i'm tired though lol

ttung replied:
It's add an accession for this entity.  How about "create_for_entity"?  It's not exactly accurate either because it's not just creating it, it's attaching it to the entity.

phoenixAja replied:
what about `attach_to_entity`? for some reason that seems more clear to me

ttung replied:
Updated.

Commit:
Convert accessions to use a workflow model

### Description
This turned out to be a lot more complicated than I expected.  Most of the complexity lies in the conversion logic (which we are **not** doing) and the logic for sample view.  The latter is required and now adds another roundtrip to the server.  I suspect it should be possible to do what we want in one query, but I can't seem to massage sqlalchemy to do what I want.

Rest of this is pretty straightforward.

#### Issue
[ch123630](https://app.clubhouse.io/genepi/story/123630/move-accessions-to-a-workflow-model)

### Test plan
make -j style unit-tests
